### ISSUE TITLE
[VL] EP: Memory management enhancements after integrating with Velox arbitrator API

### ISSUE THREAD

**Issue by zhztheplayer** [oap-project/gluten#2849]
### Description

(`EP` in title stands for **enhancement proposal**. Let's have a standardized way to organize all the EPs in future.)

This is an umbrella issue for recent changes and other planned changes, for better integration with Velox arbitrator API.

## Background

We migrated to Velox arbitrator API since PR #2321, and introduced [ListenableArbitrator](https://github.com/oap-project/gluten/commit/054de996f0b6b6526d82d7387ede85141c228ec5#diff-163aa1d4bcffd7b4dd3b270eab81f55748d7e44fe185cb9a8c736f8f20b2d3ccR99) to report Velox's memory usage back to Spark memory manager.

However the new design would increase the risk of getting OOM from running since Velox's arbitrators operate on the memory pools based on their **capacity** rather than the actual allocated memory.

**capacity** in Velox memory pool is designed as a memory reservation counter that never gets shrunken unless reclaim/spill is requested, so the number can be easily larger than the actual `currentBytes()` which is the real RAM memory usage of the program. Unfortunately in Gluten we create a lot of **temporary** Velox memory pools that however has long lifecycle times (typically, closed when Spark task ends). In the arbitrator speaking, this can be risky because the held off-heap numbers of these pools are not given back to Spark until it's destructed or requested spilling, although the pools already gave back the actual allocations to OS probably.

TL;DR:
Because of the current design, Some of Velox's memory allocations were released to OS, but not given back to Spark. Thus OOMs happen.

## Solution

As the issue background explained above, we are going to solve the issue within the following directions (these directions are not exclusive to each other, although some of them could solve the OOMs):

1. Limit the number of Velox memory pools created in Gluten;
2. Shorten the lifecycle of Velox memory pools, e.g. close when task ends -> close when not used;
3. Make created Velox memory pools shrinkable (which means, the capacities can be reclaimed by Spark memory manager), currently only the ones from plan executions are shrinkable;
4. Try getting Velox's actual allocation, rather than pools' capacities which can be "fake" on top of the real RSS.

## Non-goals

It's not the goal of this proposal to deal with any form of memory leaks.

## Issues and PRs

1. https://github.com/oap-project/gluten/pull/2786
2. https://github.com/oap-project/gluten/issues/2803
    https://github.com/oap-project/gluten/pull/2802
3. https://github.com/oap-project/gluten/pull/2843
4. https://github.com/oap-project/gluten/issues/2808
    https://github.com/oap-project/gluten/pull/3014
6. https://github.com/oap-project/gluten/issues/2848

**Comment by Yohahaha** [comment]
solution 3 make more sense to me, In Spark, each memory consumer could spilled by other or itself.

**Comment by zhztheplayer** [comment]
Yes. And the listed are possible directions to solve the issue so they are not exclusive. We can fix them all to make sure this part of code robust enough.

**Comment by zhouyuan** [comment]
@zhztheplayer thanks for the write up, so the issue is there's mismatch between `capacity` and `currentBytes()` and gluten reported the wrong one to Spark, #4 seems a very straightforward solution.

**Comment by FelixYBW** [comment]
we should start to propose APIs in Velox community for #4

**Comment by FelixYBW** [comment]
Add another issue that we pre-allocate 30% more memory when velox memory pool allocate memory. Such estimation is not scalable to user case. We need to think a way to better handle this.
